src/transaction/log_replication_atomic.cpp

```diff
@@ -0,0 +1,173 @@
+		    thread_entry, m_redo_lsa);
+	    break;
+	  case LOG_DUMMY_HA_SERVER_STATE:
+	    calculate_replication_delay_or_dispatch_async<log_rec_ha_server_state> (
+		    thread_entry, m_redo_lsa);
+	    break;
+	  case LOG_TRANTABLE_SNAPSHOT:
+	    // save the LSA of the last transaction table snapshot that can be found in the log
+	    // only needed on the passive transaction server
+	    m_most_recent_trantable_snapshot_lsa.store (m_redo_lsa);
+	    break;
+	  case LOG_START_ATOMIC_REPL:
+	  case LOG_SYSOP_ATOMIC_START:
+	    if (m_atomic_helper.is_part_of_atomic_replication (header.trid))
+	      {
+		// nested atomic replication
+		assert (false);
+	      }
+	    m_atomic_helper.start_new_atomic_replication_sequence (header.trid);
+	    break;
+	  case LOG_END_ATOMIC_REPL:
+	  case LOG_SYSOP_END:
+	    if (!m_atomic_helper.is_part_of_atomic_replication (header.trid))
+	      {
+		//log here for end without start
+		assert (false);
+	      }
+	    m_atomic_helper.unfix_atomic_replication_sequence (&thread_entry, header.trid);
+	    break;
+	  default:
+	    // do nothing
+	    break;
+	  }
+
+	{
+	  std::unique_lock<std::mutex> lock (m_redo_lsa_mutex);
+
+	  // better to be checked as soon as possible during the processing loop
+	  // however, this would need one more mutex lock; therefore, suffice to do it here
+	  assert (m_replication_active);
+
+	  m_redo_lsa = header.forw_lsa;
+	}
+
+	// to accurately track progress and avoid clients to wait for too long, notify each change
+	m_redo_lsa_condvar.notify_all ();
+
+	m_perfmon_redo_sync.track_and_start ();
+      }
+  }
+
+  template <typename T>
+  void
+  atomic_replicator::read_and_redo_record_with_atomic_consideration (cubthread::entry &thread_entry, LOG_RECTYPE rectype,
+      const log_lsa &prev_rec_lsa, const log_lsa &rec_lsa, TRANID trid)
+  {
+    m_redo_context.m_reader.advance_when_does_not_fit (sizeof (T));
+    const log_rv_redo_rec_info<T> record_info (rec_lsa, rectype,
+	m_redo_context.m_reader.reinterpret_copy_and_add_align<T> ());
+
+    // only mvccids that pertain to redo's are processed here
+    const MVCCID mvccid = log_rv_get_log_rec_mvccid (record_info.m_logrec);
+    assert_correct_mvccid (record_info.m_logrec, mvccid);
+    log_replication_update_header_mvcc_vacuum_info (mvccid, prev_rec_lsa, rec_lsa, m_bookkeep_mvcc_vacuum_info);
+
+    // Redo b-tree stats differs from what the recovery usually does. Get the recovery index before deciding how to
+    // proceed.
+    const LOG_RCVINDEX rcvindex = log_rv_get_log_rec_data (record_info.m_logrec).rcvindex;
+    const VPID log_vpid = log_rv_get_log_rec_vpid<T> (record_info.m_logrec);
+    if (m_atomic_helper.is_part_of_atomic_replication (trid))
+      {
+	m_atomic_helper.add_atomic_replication_unit (&thread_entry, trid, rec_lsa, rcvindex, log_vpid, m_redo_context,
+	    record_info);
+      }
+
+    if (rcvindex == RVBT_LOG_GLOBAL_UNIQUE_STATS_COMMIT)
+      {
+	read_and_redo_btree_stats (thread_entry, record_info);
```

Comment:
what if this log record - which has a particular handling - is also part of an atomic replication sequence? will that be handled correctly (if at all)?